design/namespaced-roles.md

```diff
@@ -0,0 +1,299 @@
+* **Status**: Draft #1
+* **JIRA**: TBD
+* **Github Discussion**: [Namespaced Roles](https://github.com/keycloak/keycloak/discussions/8516)
+
+## Motivation
+
+In Keycloak, roles are defined at the Realm and Client levels. These roles can then be assigned directly to a User or to a Group, making every user in that group inherit the role.
+
+This is done through role mapping tables in the database, limiting the way you can assign roles to those two use cases.
+
+We will need more flexibility in the future, for example, to assign a role to a user in the context of a specific group (instead of having everyone in the group inherit a role) or, as the tenancy model gets implemented, a way to both define and assign roles in the context of a tenant.
+
+In order to have this flexibility and still support current use cases, we could create a new kind of role or backport existing roles to this new format: Namespaced roles.
+
+## Specification
+
+A Namespaced Role would follow a certain structure that would unambiguously identify it across the system, following a hierarchical structure.
+
+They would work similarly to ARNs in AWS, where a portion of the format is fixed and always identifiable by position, with the possibility of containing empty fields, containing a final variable part that could be freely named.
+
+An example would be:
+
+`role_v1:/\<realm\>/\<tenant\>/\<client\>/\<group\>/<roleName>`
```

Comment:
In my mind a role namespace was following the syntax:

[namespace/]role-name

```
kc/clients/myclient/myrole
```

Reply:
This follows the same format, but the namespace is composed of a fully identifiable path, made up by the scopes where the role is created.

This is to make it possible to decompose a role by its namespace and extract variables from it and to make querying them easy.

Reply:
Yeah, I think I agree with having the role_name first is better.